src/engine.ml

```diff
@@ -535,16 +567,18 @@ let unpad block_size cs =
   else
     Error (`Msg "bad padding")
 
-let data_out (ctx : keys) compress protocol rng key data =
-  (* output is: packed_id 0xfa data, then wrap openvpn partial header
+let out ?ts (ctx : keys) compress rng data =
```

Comment:
yeah, so in pre-shared static key mode (`secret yyy`), the wire layout is:
`<packet id> <timestamp> <data>`

while in tls (forward-secure) mode the layout is:
`<packet id> <data>`

the reason is to limit replay attacks as far as I understand. since the wire format and encryption is similar enough, I think it makes sense to use the same function. The timestamp is actually a unix time_t (i.e. seconds since epoch) as 32 bit integer (yuck, this will wrap in 2038). Now, the decode / decrypt and encode / encrypt functions do not take the entire `State.t` (which would allow a match and use the `t.now` directly), but only the components needed. I agree that the above API is fragile and should be better documented (luckily it is not exposed, but only internal).

Reply:
done in f06621f